tests/subsys/usb_c/mocks/mock_tester.c

```diff
@@ -0,0 +1,393 @@
+/*
+ * Copyright 2023 The Chromium OS Authors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <string.h>
+#include <stddef.h>
+#include <zephyr/drivers/usb_c/usbc_pd.h>
+
+#include "mock_tester.h"
+#include "mock_tcpc.h"
+
+/**
+ * @breif Tester data
+ */
+struct mock_tester_t {
+	/* pointer to mock tcpc */
+	struct mock_tcpc_data *tcpc;
+	/* CC1 voltage state */
+	enum tc_cc_voltage_state cc1;
+	/* CC2 voltage state */
+	enum tc_cc_voltage_state cc2;
+	/* CC Pull set by UUT */
+	enum tc_cc_pull uut_cc_pull;
+	/* CC Polarity set by UUT */
+	enum tc_cc_polarity uut_polarity;
+
+	/* VBUS set by UUT or Tester */
+	int vbus;
+	/* PD Message to send to the UUT */
+	struct pd_msg tx_msg;
+	/* PD Message received from the UUT */
+	struct pd_msg rx_msg;
+
+	/* PD message ID */
+	int msg_id;
+	/* PD revision */
+	enum pd_rev_type rev;
+	/* Tester's Power Role */
+	enum tc_power_role power_role;
+	/* Tester's Data Role */
+	enum tc_data_role data_role;
+};
+
+/* Test data object */
+static struct mock_tester_t tester;
+
+/**
+ * @brief Set the PD message ID
+ *
+ * @param id The id to set
+ */
+void tester_msgid_set(int id)
+{
+	tester.tx_msg.header.message_id = id & 7;
+}
+
+/**
+ * @brief Increment the PD message ID
+ */
+void tester_msgid_inc(void)
+{
+	tester.tx_msg.header.message_id++;
+	tester.tx_msg.header.message_id &= 7;
+}
+
+/**
+ * @brief Get the PD message ID
+ *
+ * @retval message id
+ */
+int tester_get_msgid(void)
+{
+	return tester.tx_msg.header.message_id;
+}
+
+/**
+ * @brief Send a PD message to the UUT
+ *
+ * @param msg PD message to send
+ */
+void tester_transmit_data(struct pd_msg *msg)
+{
+	if (tester.tcpc->rx_enable) {
+		memcpy(&tester.tcpc->rx_msg, msg, sizeof(struct pd_msg));
+		tester.tcpc->pending_rx_msg = true;
+	}
+}
+
+/**
+ * @brief Send a Hard Reset to the UUT
+ */
+void tester_send_hard_reset(void)
+{
+	memset(&tester.tcpc->rx_msg, 0, sizeof(struct pd_msg));
+	tester.tcpc->rx_msg.type = PD_PACKET_TX_HARD_RESET;
+	tester.tcpc->pending_rx_msg = true;
+}
+
+/**
+ * @brief Send a Control Message to the UUT
+ *
+ * @param type Control message type
+ * @param true if the message ID should be incremented
+ */
+void tester_send_ctrl_msg(enum pd_ctrl_msg_type type, bool inc_msgid)
+{
+	tester.tx_msg.type = PD_PACKET_SOP;
+	tester.tx_msg.header.message_type = type;
+	tester.tx_msg.header.port_data_role = tester.data_role;
+	tester.tx_msg.header.specification_revision = tester.rev;
+	tester.tx_msg.header.port_power_role = tester.power_role;
+
+	if (inc_msgid) {
+		tester_msgid_inc();
+	}
+
+	tester.tx_msg.header.number_of_data_objects = 0;
+	tester.tx_msg.header.extended = 0;
+	tester.tx_msg.len = 0;
+
+	tester_transmit_data(&tester.tx_msg);
+}
+
+/**
+ * @brief Send a Data Message to the UUT
+ *
+ * @param type Data message type
```

Comment:
What about `len`? (Nit: I would use something like `num_objects`. `tester.tx_msg.len` is evidently denominated in bytes, whereas `len` is denominated in 4-byte PD data objects.)